### ISSUE TITLE
Allow config files to have dependencies 

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#4188]
From @quicksketch in https://github.com/backdrop/backdrop-issues/issues/1811

I love the idea of importing a whole site into a fresh environment, and having it automatically download and install the needed modules. That would be slick! But first, let's fix this part so that config files know which modules they need.

This also makes me wonder if the config files should have a list of dependencies, rather than a single module. e.g. a `_config_dependencies` entry rather than just `_config_module`. Then module extending config files (e.g. the `node.type.*` config files) could do something like:

```php
$config = config('node.type.foo');
$config->set('my_module_setting', 'foo');
$config->addDependency('my_module');
$config->save();
```

Then importing the content type config file would know it needs not just the node module but also the extending module(s). Right now we don't actually track this at all, so we have no idea that a module that extends an existing config file would also be needed.

**Comment by docwilmot** [comment]
I'm trying to understand when this would be required. I imagine we mean when a module saves data to a config file (if that module does not own the `hook_config_info()` for that config file) it should add itself as a dependency?

**Comment by jenlampton** [comment]
Yeah, something like that. Maybe path would add's itself to a dependency on node, as soon as a node type gets a default path pattern?

**Comment by stpaultim** [comment]
I don't yet know if this is a good idea, but the use case that occurs to me for config files declaring dependencies would be if we allow batching config file into recipes, ala:nhttps://github.com/backdrop/backdrop-issues/issues/3763

I created an experimental project, which is basically a library of config recipes to add features to a site. https://github.com/backdrop-contrib/config_recipes

These recipes include the config for a content types, custom fields and field instances, as well as views. An example recipe that is included is ```locations.``` The locations recipes requires that the address field module and the geo suite of modules are installed. For now, I just included a note in the readme for that recipe.

This use case assume that the Backdrop equivalent for the features module involves packages of config files. Other might have better ideas for accomplishing this goal.

**Comment by stpaultim** [comment]
I uploaded a video demo of a config recipe use case for this: https://youtu.be/VgFibVrEZEw

**Comment by klonos** [comment]
Here's a recent use case where this is needed: #6639 ...exporting a layout config file without the respective menu config file and then importing that config to the target site causes errors.
